sigmastate/src/main/scala/org/ergoplatform/ErgoLikeContext.scala

```diff
@@ -111,6 +111,18 @@ class ErgoLikeContext(val lastBlockUtxoRoot: AvlTreeData,
   */
   val self: ErgoBox = boxesToSpend(selfIndex)
 
+  /** Current version of the ErgoTree executed by the interpreter.
+    * This property is used to implement version dependent operations and passed to
+    * interpreter via [[specia.sigma.Context]].
+    * The value cannot be assigned on [[ErgoLikeContext]] construction and must be
```

Comment:
Why can't the property be passed via constructor. 

Initialization order is error-prone (and doing 2x allocations) , as clients getting the new version of the code will get errors in runtime only.

Reply:
1) The overhead of allocation is negligible since it only 1 allocation per script and we don't have even thousands of scripts per second (it would only make sense for millions per second)
For comparison, every ErgoTree operation in v5.0 interpreter creates a new object (because of immutability), so at least 1 allocation per operation. 
2) This context parameter is only necessary inside interpreter and can be safely initialized based on the actually passed ErgoTree. So this is encapsulated into Interpreter and the client don't have to deal with this. (the motivation is similar to evaluationMode we discussed before an you were in favor of encapsulation).
By default it is None and the CostingDataContext, which is created for each ErgoLikeContext and actually used in the reduction have this  `currentErgoTreeVersion: Byte` i.e. it is required.
There is no way to execute the interpreter without setting this parameter.

3) If make it a context parameter it will require a lot of changes everywhere and each client will need to take care of passing the correct version for the given ErgoTree. In particular need to know how to extract that version.

Reply:
In Ergo node, there are few instantiations of ErgoLikeContext, and also 

```
class ErgoContext(val stateContext: ErgoStateContext,
                  transactionContext: TransactionContext,
                  inputContext: InputContext,
                  override val costLimit: Long,
                  override val initCost: Long)
  extends ErgoLikeContext(ErgoInterpreter.avlTreeFromDigest(stateContext.previousStateDigest),
    stateContext.sigmaLastHeaders,
    stateContext.sigmaPreHeader,
    transactionContext.dataBoxes,
    transactionContext.boxesToSpend,
    transactionContext.spendingTransaction,
    inputContext.selfIndex,
    inputContext.extension,
    stateContext.validationSettings.sigmaSettings,
    costLimit,
    initCost,
    activatedScriptVersion = (stateContext.blockVersion - 1).toByte // block version N of ErgoProtocol corresponds to version N-1 of ErgoTree (aka script version)
  )
``` 

not sure about other applications.

ErgoContext will implicitly get the iini problems, for example, just with dependency update. And the problem will appear in runtime only. That's not how a library should be developed.